### PR TITLE
Monorepo

### PR DESCRIPTION
**Pull request by HenriBeck**
Started on converting this repo into a monorepo.

Some things to discuss:
- Fixed or independent version mode?
- Which repositories to import?
- Moving to an NPM org as well? (like babel)

#714

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,142 +1,5 @@
 {
-  "name": "jss",
```
Comment (TrySound) [inline package.json]: Keep the `name` and add `private: false`.

**Comment by TrySound** [comment]
I'm thinking maybe it will be better to move all plugins first and provide them build step and then work with `jss` itself. Because with current strategy we move a lot of stuff deeper and then we will a lot of stuff hoist in the root.

**Comment by TrySound** [comment]
Also it would be good to move each plugin in separate PRs so we could review faster.

**Comment by TrySound** [comment]
Also we probably should not change package names and publish one patch with new links so nobody tried to look at old repos.

**Comment by HenriBeck** [comment]
I would rather move `jss` in this pr and setup build and test scripts in this pr and then move the other packages in separate PRs.

@TrySound so you are against switching to an npm org?

File: package.json [hunk 2]
```diff
@@ -1,142 +1,5 @@
...
-  "description": "A lib for generating Style Sheets with JavaScript.",
-  "version": "9.8.1",
-  "author": {
-    "name": "Oleg Slobodskoi",
-    "email": "oleg008@gmail.com"
-  },
-  "repository": {
```
Comment (TrySound) [inline package.json]: `author` and `repository` should probably also stay here

File: package.json [hunk 3]
```diff
@@ -1,142 +1,5 @@
...
-    "css",
-    "components",
-    "composable",
-    "css in js",
-    "css-in-js"
-  ],
-  "license": "MIT",
-  "main": "./lib/index.js",
-  "engines": {
-    "node": ">=4"
```
Comment (TrySound) [inline package.json]: This should stay here to validate devDependencies

File: package.json [hunk 4]
```diff
@@ -1,142 +1,5 @@
...
-  },
-  "devDependencies": {
```
Comment (TrySound) [inline package.json]: All development stuff should stay here. It's cleaner when packages have only production dependencies.

File: package.json [hunk 5]
```diff
@@ -1,142 +1,5 @@
...
-  "files": [
-    "dist",
-    "lib",
-    "src",
-    "flow-typed",
-    "readme.md",
-    "changelog.md",
-    "LICENSE"
-  ],
-  "lint-staged": {
```
Comment (TrySound) [inline package.json]: `lint-staged` should also be run from the root

**Comment by kof** [comment]
btw. we don't own jss namespace on npm, we do https://www.npmjs.com/~cssinjs though

File: package.json (hunk 5 above)
Comment (HenriBeck) [reply package.json]: Yeah, I still need to move all the dev deps and setup scripts in the root package.

**Comment by kof** [comment]
Wait, we don't need the user name to publish under @jss scope

**Comment by kof** [comment]
Just looked at babel, they are not using the namespace. What are the pros/cons?

**Comment by HenriBeck** [comment]
https://www.npmjs.com/docs/orgs/migrating-a-user-account.html I think thats the problem, there is already a user with the jss name.

Babel switched to namespaces with v7

**Comment by kof** [comment]
I asked the person, lets think of alternatives.

**Comment by HenriBeck** [comment]
Well, we don't have to switch, it was more of an idea. I think it just unifies the packages a bit more.

**Comment by kof** [comment]
We can migrate with current package names first, then if we get a good namespace we can change that as a separate step. Lets focus on things which are necessary

**Comment by kof** [comment]
@HenriBeck we got some conficts and the comments from @TrySound

**Comment by HenriBeck** [comment]
A few things about the monorepo:

- The packages will have their own build scripts
- The packages will have their own test scripts as those might differ
- I moved some configs like the rollup config into the root folder and made created a config function to be used by the packages
- Everything else lives in the root folder now (lint etc.)

Some problems which still exists for some reason:
- Babel always throws an error, this makes the test and the build scripts fail

**Comment by kof** [comment]
What if we unify all the build stuff and put only compatible repos into this monorepo? For e.g. jss and plugins only for now?

**Comment by kof** [comment]
I have the idea to move to the same testing library everywhere, but for this we need to extract browser specific part into a separate library which can keep the current complex setup with karma.

**Comment by kof** [comment]
the thing is if we keep each package with its own build, there is not much value in monorepo, I think the big value is to reduce maintenance cost of updating each repo, so we need a unification.

wdyt?

**Comment by TrySound** [comment]
You may keep individual builds. I can hoist them in the next PR then.

**Comment by kof** [comment]
lets do it before the merge @TrySound you can commit directly to this branch as well

**Comment by TrySound** [comment]
I just would like to do atomic pull requests to simplify review and see the state of work.

**Comment by kof** [comment]
that is true, unless master gets into a state where we can't release from, also you need to know that cssinjs.org website is dynamically fetching markdown files from each repo and it will need to be fixed too

**Comment by kof** [comment]
So if we have these points, we can merge:
- Can continue releasing JSS core
- Can continue releasing plugins
- Didn't break the website, can continue changing text
- Didn't break the CI

**Comment by HenriBeck** [comment]
About the build step, currently, each package has its own build npm script. We could do an approach like storybook did with having js script files which we can simply call, but each package would still need an npm script.

I've already extracted the configuration for the rollup config into a function so the different plugins have a unified config, the only things that change between them are the names.

**Comment by kof** [comment]
Yeah, its a good first step I think

File: .flowconfig [hunk 6]
```diff
@@ -3,6 +3,7 @@
 [include]
 
 [libs]
+packages/jss/flow-typed
```
Comment (TrySound) [inline .flowconfig]: already removed

File: .eslintignore [hunk 7]
```diff
@@ -1 +1,4 @@
-flow-typed/
+packages/jss/flow-typed/
```
Comment (TrySound) [inline .eslintignore]: already removed

**Comment by TrySound** [comment]
@HenriBeck Could you rebase so we could focus on actual changes?

**Comment by HenriBeck** [comment]
Yeah will rebase and finish work on this tomorrow

**Comment by HenriBeck** [comment]
As there are a lot of changes here, which when rebasing would completely revert my changes so far, I decided to start clean on a new branch too make commits a bit tidier

**Comment by HenriBeck** [comment]
#729

**Pull request closed without merging**